### PR TITLE
Autodetect msvc versions

### PR DESCRIPTION
**Pull request by KonTy**
**Changes proposed:**

-  
-  
-  

**Target branch(es):** 3.3.5/master

- [X] 3.3.5
- [X] master

**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:** (Does it build, tested in-game, etc.)

**Known issues and TODO list:** (add/remove lines as needed)

### CODE REVIEW

File: dep/boost/CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@
 #
 # This program is distributed in the hope that it will be useful, but
 # WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
-# implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. "lib${PLATFORM}-msvc*"
```
Comment (Aokromes) [inline dep/boost/CMakeLists.txt]: i suppose this is unintended change

**Comment by Rochet2** [comment]
What happens if there are multiple boost installed like this?
I _think_ it would choose completely wrong boost, like 8.0
explorer_2017-03-24_21-13-25

**Comment by tkrokli** [comment]
@Rochet2 : isn't Boost selected by checking the Windows Environment Variable `BOOST_ROOT` ?
That value will usually be set manually by the user to match the desired CMake & VS setup.

File: dep/boost/CMakeLists.txt (hunk 1 above)
Comment (KonTy) [inline dep/boost/CMakeLists.txt]: Yes this was unintended thanks for catching

**Comment by KonTy** [comment]
@Rochet2 what is the "right" boost version in your mind? I will chose the first boost it sees, how do you think we should fix it? We could have a wild card include 14 like this: lib${PLATFORM}-msvc.14* 
but this will miss next versions of boost 15,16 etc.

**Comment by Aokromes** [comment]
what Rochet2 mentions remind me when we tried to do the same for mysql, the result, a lot of people where unable to compile because auto selected mysql connector instead mysql.

**Comment by KonTy** [comment]
@Aokromes in this case we should fix the docs, because here it says it is recommended to use boost 64
https://trinitycore.atlassian.net/wiki/display/tc/Windows+Requirements 
"1.63.0 is the minimum version required, but version 1.64.0 is recommended (fully supports Visual Studio 2017)"

but the script doesn't support 64 it will not compile with 64 installed because 64 uses "lib32-msvc-14.1" not lib32-msvc-14.0.

There has to be a better way...

**Comment by Aokromes** [comment]
or fix the code :P

**Comment by ariel-** [comment]
3049d19b53db7e7f2042ee837771f7a71ad31373 ?

**Comment by Rochet2** [comment]
> @Rochet2 : isn't Boost selected by checking the Windows Environment Variable BOOST_ROOT ?
> That value will usually be set manually by the user to match the desired CMake & VS setup.

That only sets the root, those folders are inside the root. The boost root will only show where boost is installed, not which boost version to use.

> @Rochet2 what is the "right" boost version in your mind? I will chose the first boost it sees, how do you think we should fix it? We could have a wild card include 14 like this: lib${PLATFORM}-msvc.14*

The one right for the compiler. IE 14.1 if 8.0 and 14.0 and 14.1 are installed.

> 3049d19 ?

@ariel- I _guess_ that fixed the problem for 3.3.5 but isnt applied to master yet?
I wonder how it works with multiple libraries and compilers selected like that.

**Comment by jackpoz** [comment]
where's the PR description ?

**Comment by KonTy** [comment]
@jackpoz it's in the title there isn't much more to add to that.
@ariel- 
> 3049d19 I personally do not like that solution because it is hardcoding the versions.
perhaps for 2017 we can pick a range? ex: <= 14.0 

here is what I did and tested it extensively

`list(LENGTH MSVC_LIB_DIRECTORIES MSVC_LIB_DIRECTORIES_LENGTH)
MATH(EXPR MSVC_LAST_ELEMENT "${MSVC_LIB_DIRECTORIES_LENGTH}-1")
list(SORT MSVC_LIB_DIRECTORIES)
list(GET MSVC_LIB_DIRECTORIES ${MSVC_LAST_ELEMENT} BOOST_LIBRARYDIR)`

**Comment by DDuarte** [comment]
What's the behavior of the official FindBoost.cmake on this topic?

**Comment by ariel-** [comment]
> What's the behavior of the official FindBoost.cmake on this topic?

The one I ported in 3049d19b53db7e7f2042ee837771f7a71ad31373, it's TC that hardcodes and "auto selects" best version here: https://github.com/TrinityCore/TrinityCore/blob/3.3.5/dep/boost/CMakeLists.txt#L11-L33

**Comment by joschiwald** [comment]
> What's the behavior of the official FindBoost.cmake on this topic?
yeah we should update our FindBoost.cmake, many stuff changed since last 2 years
for me, it is working well

**Comment by Aokromes** [comment]
Conflicting files
dep/boost/CMakeLists.txt

**Comment by Aokromes** [comment]
why close?

**Pull request closed without merging**